feat: FIT-610: FSM core model states setup

bmartel opened the pull request:
This pull request introduces a foundational set of models and supporting code for tracking state transitions (finite state machine, FSM) for Label Studio core entities: Tasks, Annotations, and Projects. It adds new migration files, core state models, state choice enums, and improves the state manager interface for clarity and extensibility.

The most important changes are:

**Core FSM Models and Schema:**
- Added new Django models `TaskState`, `AnnotationState`, and `ProjectState` in `fsm/models.py` to track state transitions for their respective entities, including denormalized fields for performance and index optimizations. These models are registered for FSM use and support a simple 3-state workflow for each entity.
- Introduced the initial migration (`0001_initial.py`) that creates the above state models with all necessary fields, relationships, and indexes to support efficient querying and reporting.

**State Choices Registry:**
- Added `state_choices.py`, defining `TaskStateChoices`, `AnnotationStateChoices`, and `ProjectStateChoices` as Django `TextChoices` enums, and registered them for FSM usage. This centralizes and standardizes the possible states for each entity type.

**State Manager API Improvements:**
- Renamed the core state retrieval method from `get_current_state` to `get_current_state_value` for clarity, and provided a backward-compatible alias. This makes the API clearer and future-proof. [[1]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L62-R62) [[2]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L76-R76) [[3]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L103-R103) [[4]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L130-R135)

**Imports and Registry Integration:**
- Updated imports in `fsm/models.py` to use the new registry system for state models and choices, ensuring all FSM components are properly registered and discoverable.

**Improved Test Coverage**
- Added more test coverage for the additional code as well as missing lines from the previous PR.

bmartel committed:
feat: FIT-587: FSM Architecture setup

bmartel committed:
include organization on base state model

bmartel committed:
refactoring core definitions

bmartel committed:
define core entity states in full

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
fixed imports

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
fix imports

bmartel committed:
fix imports

bmartel committed:
fix fsm tests

bmartel committed:
fix fsm tests

bmartel committed:
fix linting errors

bmartel committed:
fix fsm concurrency tests

bmartel committed:
fix remaining fsm tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
updating doc strings

bmartel committed:
fixing registry

bmartel committed:
fixing registry

bmartel committed:
fixing tests

bmartel committed:
updating README

bmartel committed:
Apply suggestion from @bmartel

bmartel committed:
avoiding import cycle due to types

bmartel committed:
feat: FIT-587: FSM core model setup

bmartel committed:
fix imports

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fix imports

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fix typing and decorator definition for state model

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
adding migrations

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
moving settings from extensions to settings file

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
moving the API integration FSM out of this PR

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
updating readme

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
renaming transition decorator for consistency

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fixing tests to align with refactors and renames of transition functions

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing serializers from this PR, will be included in a different one

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
Update label_studio/fsm/state_manager.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
updating based on feedback about test assertions/file names

bmartel committed:
updating based on feedback about test file names

bmartel committed:
consolidate transition util get_valid_states to being the same function get_available_states

bmartel committed:
improving transition handling

bmartel committed:
improving transition handling by consolidating everything to use a single entry point

bmartel committed:
ensure write through cache sets on_commit of the transaction

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
type fixes

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
update tests to ensure coverage

bmartel committed:
fixing lint errors

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
reverting poetry.lock to develop to regenerate it on CI

bmartel committed:
Merge branch 'fb-fit-587-fsm' of github.com:HumanSignal/label-studio into fb-fit-587-fsm

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446529166

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
updating docs

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
removing unused code

bmartel committed:
applying feedback from CR

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fixing tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
structure the logging so it can be made useful for metrics in grafana

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fix tests

bmartel committed:
fix tests

bmartel committed:
fix implementation

bmartel committed:
fix implementation

bmartel committed:
Merge branch 'develop' into fb-fit-587-fsm

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fix initial migration

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
refactor code to remove function imports

bmartel committed:
refactor code to remove redundant code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
remove no longer required test

bmartel committed:
updating docs

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fixing lint errors

bmartel committed:
fixing lint errors

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-610-fsm-models

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17554612979)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17554612979

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17554612979

bmartel committed:
fixing tests for windows SQLITE

bmartel committed:
Merge branch 'fb-fit-610-fsm-models' of github.com:HumanSignal/label-studio into fb-fit-610-fsm-models

mcanu commented on label_studio/fsm/models.py:
```diff
@@ -194,3 +196,148 @@ def _get_entity_field_name(cls) -> str:
+    """
+
+    # Entity Relationship
+    task = models.ForeignKey('tasks.Task', related_name='fsm_states', on_delete=models.CASCADE, db_index=True)
```
https://docs.djangoproject.com/en/5.2/ref/models/fields/#foreignkey db index is already created for foreign keys

bmartel replied:
Thanks, I had a feeling that would likely be the default behaviour, will leave it be default then as to not be overly redundant.

mcanu commented on label_studio/fsm/models.py:
```diff
@@ -194,3 +196,148 @@ def _get_entity_field_name(cls) -> str:
+            # Critical: Latest state lookup (current state determined by latest UUID7 id)
+            # Index with DESC order explicitly supports ORDER BY id DESC queries
+            models.Index(fields=['task_id', '-id'], name='task_current_state_idx'),
+            # Reporting and filtering
+            models.Index(fields=['project_id', 'state', '-id'], name='task_project_state_idx'),
+            models.Index(fields=['organization_id', 'state', '-id'], name='task_org_reporting_idx'),
+            # History queries
+            models.Index(fields=['task_id', 'id'], name='task_history_idx'),
```
Any reason to explicitly name the indexes and not let Django name them automatically?

bmartel replied:
Old habit from other ORM's/languages. I tend to like to be able to search for the indices and see exactly that in both code and in the database as well.

mcanu commented on label_studio/fsm/models.py:
```diff
@@ -194,3 +196,148 @@ def _get_entity_field_name(cls) -> str:
+        return {
+            'task_id': entity.task.id,
+            'project_id': entity.task.project_id,
+            'completed_by_id': entity.completed_by.id if entity.completed_by else None,
```
```suggestion
            'completed_by_id': entity.completed_by_id if entity.completed_by_id else None,
```
This saves a query to get the user

mcanu commented on label_studio/fsm/models.py:
```diff
@@ -194,3 +196,148 @@ def _get_entity_field_name(cls) -> str:
+    def get_denormalized_fields(cls, entity):
+        """Get denormalized fields for ProjectState creation"""
+        return {
+            'created_by_id': entity.created_by.id if entity.created_by else None,
```
```suggestion
            'created_by_id': entity.created_by_id if entity.created_by_id else None,
```

mcanu commented on label_studio/fsm/state_manager.py:
```diff
@@ -127,7 +127,12 @@ def get_current_state(cls, entity: Model) -> Optional[str]:
+    @classmethod
+    def get_current_state(cls, entity: Model) -> Optional[str]:
+        """Backward compatibility method - calls get_current_state_value"""
+        return cls.get_current_state_value(entity)
```
Why do we need to support a Backward compatibility method? Just trying not to introduce tech debt from the beginning. Lets just replace the usage.

mcanu commented on label_studio/fsm/tests/test_fsm_integration.py:
```diff
@@ -0,0 +1,417 @@
+    def setUp(self):
+        self.user = User.objects.create_user(email='test@example.com', password='test123')
+        self.project = Project.objects.create(title='Test Project', created_by=self.user)
+        self.task = Task.objects.create(project=self.project, data={'text': 'test'})
```
Recommendation to use Factories here

mcanu reviewed (approved):
Some small comments, LGTM

bmartel committed:
Update label_studio/fsm/models.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
Update label_studio/fsm/models.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
applying feedback to remove the additional method get_current_state and utilize get_current_state_value and get_current_state_object as the only means for retrieving current_state

bmartel committed:
removing redundant db_index=True on task fk

bmartel committed:
using factories as mentioned in PR feedback

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17585497946)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17585497946

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17585497946

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17591882567)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17591882567

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17591882567

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17615595310)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17615595310

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17615595310

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17618306507)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17618306507

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17618306507

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17627590890)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17627590890

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17627590890

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17644693280)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17644693280

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17644693280
